### PR TITLE
Allow to search files by comments

### PR DESCRIPTION
**Pull request by nickvergessen**
> bildschirmfoto von 2018-04-18 11-25-26

PS: this is an accident, I was just preparing search for chat messages in talk.

### CODE REVIEW

File: lib/private/Comments/Manager.php [hunk 1]
```diff
@@ -376,6 +376,48 @@ public function getForObject(
...
+	/**
+	 * Search for comments with a given content
+	 *
+	 * @param string $search content to search for
+	 * @param string $objectType Limit the search by object type
+	 * @param string $objectId Limit the search by object id
+	 * @param string $verb Limit the verb of the comment
+	 * @return IComment[]
+	 */
+	public function search(string $search, string $objectType, string $objectId, string $verb): array {
```
Comment (nickvergessen) [inline lib/private/Comments/Manager.php]: I guess we should add an limit/offset here, opinions?

File: apps/comments/lib/Search/CommentSearchResult.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+		if ((strlen($message) - $end) <= 25) {
+			$end = strlen($message);
+			$suffix = '';
+		} else {
+			$end += 25;
+			$suffix = '…';
+		}
+
+		return $prefix . substr($message, $start, $end - $start) . $suffix;
```
Comment (nickvergessen) [inline apps/comments/lib/Search/CommentSearchResult.php]: This is rather arbitrary atm, we could also search for the beginning/end of the next word and if it is within the range cut there. On the other hand I like simplicity

File: lib/private/Comments/Manager.php (hunk 1 above)
Comment (skjnldsv) [reply lib/private/Comments/Manager.php]: Can lead to pretty big requests, yes. 
Is this an admin only tool?

Comment (nickvergessen) [reply lib/private/Comments/Manager.php]: No, it's the normal search in the top right, which finds files for you :dancer:

Comment (nickvergessen) [reply lib/private/Comments/Manager.php]: The problem with pagination is, we paginate the comments. Afterwards we need to look if the user can actually see the file where the comment lives. so while we paginate, the page could still be empty, unless we loop until we found enough comments....

But offset is pretty hard with that, so maybe we don't go for offset at all.

File: apps/comments/lib/Search/CommentSearchResult.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Comments\Search;
+
+use OCP\Files\NotFoundException;
+use OCP\Search\Result;
+
+class CommentSearchResult extends Result {
```
[the remark this answers is not in the document]
Comment (blizzz) [reply apps/comments/lib/Search/CommentSearchResult.php]: could be imported under an alias or using full name space. anyway, not that important.

File: lib/public/Comments/ICommentsManager.php [hunk 4]
```diff
@@ -120,6 +120,18 @@ public function getForObject(
...
+	 * Search for comments with a given content
+	 *
+	 * @param string $search content to search for
+	 * @param string $objectType Limit the search by object type
+	 * @param string $objectId Limit the search by object id
+	 * @param string $verb Limit the verb of the comment
+	 * @return IComment[]
+	 * @since 14.0.0
+	 */
+	public function search(string $search, string $objectType, string $objectId, string $verb): array;
```
Comment (blizzz) [inline lib/public/Comments/ICommentsManager.php]: FakeManager in Tests needs to implement this as well

**Comment by MorrisJobke** [comment]
It doesn't find the comments for me - it also never gets into the search provider of comments when adding a breakpoint there :/

**Comment by MorrisJobke** [comment]
🏓

**Comment by nickvergessen** [comment]
Fixed searching, however the results are not looking like the screenshot anymore.
Was there a major redesign/refactoring of the search result design?

**Comment by MorrisJobke** [comment]
> Was there a major redesign/refactoring of the search result design?

Yes - #9912

**Comment by MorrisJobke** [comment]
The search itself doesn't work for me. I shared a file with another user and both commented with:

> Lorem ipsum dolor sit amet, consectetur, adipisci velit

Searching then for `Lorem` does not show any search result. It was on a file. Anything that I miss?

**Comment by nickvergessen** [comment]
I just rebased and tested this again: works pretty fine here for both the sharer and the sharee

**Comment by MorrisJobke** [comment]
Doesn't show anything for me on this branch (left is heinz, right is admin, both commented with a string that is then searched for):

bildschirmfoto 2018-07-05 um 12 00 34


What do I do wrong?

**Comment by nickvergessen** [comment]
So the entry point is https://github.com/nextcloud/server/blob/7674e591bdb8060227da3411ffd5cc43f7483f51/core/Controller/SearchController.php#L49
Once I step into that, after the init method, I can see this->providers containing the new OCA\Comments\Search\Provider:
https://github.com/nextcloud/server/blob/21a720e026368c5d1c1c55e89cd5630428a6eeb5/lib/private/Search.php#L49

And from there on the debugger should be able to step through and show you where you are leaving the trail

**Comment by MorrisJobke** [comment]
> Once I step into that, after the init method, I can see this->providers containing the new OCA\Comments\Search\Provider:

For me that was not the case 😢 

bildschirmfoto 2018-07-18 um 09 30 54

**Comment by MorrisJobke** [comment]
The constructor of the class `OCA\Comments\AppInfo\Application` is never called. I needed to actively instantiate it in the app.php of the comments app. Then the provider is properly loaded but the UI is still broken: 

bildschirmfoto 2018-07-18 um 09 39 43
bildschirmfoto 2018-07-18 um 09 39 57

**Comment by nickvergessen** [comment]
Fixed the code so the constructor is always called.

**Comment by MorrisJobke** [comment]
it still looks like that and the path and the avatar is not set properly:

bildschirmfoto 2018-07-18 um 17 51 29

**Comment by nickvergessen** [comment]
Anyone with more frontend knowledge volunteering to do that?

**Comment by danxuliu** [comment]
I will take care of this.

**Comment by danxuliu** [comment]
I have rebased on master and made the following changes:
- Moved the last commit to the first one
- Squashed the fixes into the commits that they fixed
- Added a missing type in a parameter
- Restored [the return type in the documentation of `OCP\Search\Provider::search`](https://github.com/nextcloud/server/blob/ba87db3fccb40aa58d84d12932424c83a4db411f/lib/public/Search/Provider.php#L87) from `Result[]` to `array` (of course feel free to add it back, but please modify it everywhere, not just in the base class leaving all the subclasses dangling with the old type ;-) )
- Fixed duplicated search results (for some reason, that I have not digged into, the application is created twice; as the provider was registered in the constructor it was registered twice and results were duplicated; now the provider is registered in `register()`)
- Fixed search before pagination was introduced (the parameters were passed in wrong order to `$cm->search`, and `users` should be passed to `resolveDisplayName`)
- Fixed UI (the plugin [now has to be registered](https://github.com/nextcloud/server/pull/9912) with `OCA.Search.Core` instead of `OCA.Search`)
- Renamed `renderFileResult` and `handleFileClick` to `renderCommentResult` and `handleCommentClick` (nitpicking, the name should refer to the type of the result)
- Added integration tests (no time for unit or acceptance tests, sorry ;-) )

All in all just minor changes; besides the integration tests the bulk of the pull request is still the same (and I have not modified nor tested the pagination). In any case, it has my :+1: 

Some details:
- The link to the file is not included in the search results
- The `name` property in the result includes the full comment, without ellipsis; I have no idea if that is right or not. In any case the UI shows the ellipsized version.

**Comment by danxuliu** [comment]
Ok, give me a second, I made a stupid change... :P

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
I just made the icon brighter (top before, bottom after):

bildschirmfoto 2018-07-25 um 20 28 09

**Comment by MorrisJobke** [comment]
> In any case, it has my 👍

-> Merging

**Pull request merged**